File: lib/aura/llm/http_client.rb
# frozen_string_literal: true

require "net/http"
require "json"
require "aura/errors"

module Aura
  module LLM
    class HttpClient
      # Makes a POST request and returns the parsed JSON response or yields raw chunk lines for streaming
      def self.post(url_str, headers, body_hash, timeout: 120, stream: false, &block)
        uri = URI(url_str)
        req = Net::HTTP::Post.new(uri)
        headers.each { |k, v| req[k] = v }
        req.body = JSON.dump(body_hash)

        http_opts = {
          use_ssl: uri.scheme == "https",
          open_timeout: 10,
          read_timeout: timeout,
          write_timeout: 10
        }

        begin
          Net::HTTP.start(uri.host, uri.port, http_opts) do |http|
            if stream
              http.request(req) do |res|
                validate_response_code!(res)
                res.read_body do |chunk|
                  block.call(chunk) if block_given?
                end
              end
            else
              res = http.request(req)
              validate_response_code!(res)
              begin
                JSON.parse(res.body)
              rescue JSON::ParserError => e
                raise Aura::LLMError, "Failed to parse JSON response: #{e.message}"
              end
            end
          end
        rescue Net::OpenTimeout, Net::ReadTimeout => e
          raise Aura::LLMTimeoutError, "LLM request timed out: #{e.message}"
        rescue Aura::LLMError => e
          raise e
        rescue StandardError => e
          raise Aura::LLMError, "LLM connection failed: #{e.message}"
        end
      end

      def self.validate_response_code!(response)
        code = response.respond_to?(:code) ? response.code.to_i : 200
        return if code >= 200 && code < 300

        error_message = if response.respond_to?(:body) && response.body && !response.body.to_s.strip.empty?
                          response.body.to_s.strip
                        elsif response.respond_to?(:read_body)
                          begin
                            response.read_body.to_s.strip
                          rescue StandardError
                            ""
                          end
                        else
                          ""
                        end

        if error_message.empty?
          error_message = "HTTP Status #{code}"
        else
          begin
            parsed = JSON.parse(error_message)
            error_message = parsed.dig("error", "message") || parsed["message"] || error_message
          rescue StandardError
          end
        end

        case code
        when 401, 403
          raise Aura::LLMAuthError, "Authentication failed: #{error_message}"
        when 400
          raise Aura::LLMBadRequestError, "Bad request: #{error_message}"
        when 408, 504
          raise Aura::LLMTimeoutError, "Request timed out: #{error_message}"
        when 429
          raise Aura::LLMRateLimitError, "Rate limit exceeded: #{error_message}"
        when 500..599
          raise Aura::LLMServerError, "Server error (#{code}): #{error_message}"
        else
          raise Aura::LLMError, "LLM API Error (#{code}): #{error_message}"
        end
      end
    end
  end
end

File: src/core/llm/httpClient.ts
import {
  LLMAuthError,
  LLMBadRequestError,
  LLMError,
  LLMRateLimitError,
  LLMServerError,
  LLMTimeoutError,
} from './errors.js';

export class HttpClient {
  /**
   * Makes a POST request and returns the parsed JSON response or streams chunk lines
   */
  public static async post(
    urlStr: string,
    headers: Record<string, string>,
    bodyHash: Record<string, unknown>,
    options: {
      timeout?: number;
      stream?: boolean;
      onChunk?: (chunk: string) => void;
      signal?: AbortSignal;
    } = {},
  ): Promise<Record<string, unknown> | null> {
    const timeoutSeconds = options.timeout ?? 120;
    const controller = new AbortController();
    const timeoutId = setTimeout(
      () => controller.abort(),
      timeoutSeconds * 1000,
    );

    const onAbort = () => {
      controller.abort();
    };

    if (options.signal) {
      if (options.signal.aborted) {
        controller.abort();
      } else {
        options.signal.addEventListener('abort', onAbort);
      }
    }

    try {
      const response = await fetch(urlStr, {
        method: 'POST',
        headers,
        body: JSON.stringify(bodyHash),
        signal: controller.signal,
      });

      if (!response.ok) {
        let bodyText = '';
        try {
          bodyText = await response.text();
        } catch {
          // Ignore text reading errors
        }
        HttpClient.validateResponseCode(response.status, bodyText);
      }

      if (options.stream) {
        if (!response.body) {
          throw new LLMError('Response body is null, cannot stream');
        }

        const reader = response.body.getReader();
        const decoder = new TextDecoder();

        while (true) {
          const { done, value } = await reader.read();
          if (done) break;

          const chunk = decoder.decode(value, { stream: true });
          if (options.onChunk) {
            options.onChunk(chunk);
          }
        }
        return null;
      } else {
        const json = await response.json();
        return json as Record<string, unknown>;
      }
    } catch (err: unknown) {
      if (err instanceof Error && err.name === 'AbortError') {
        if (options.signal?.aborted) {
          throw new LLMError(`LLM request aborted: ${err.message}`);
        }
        throw new LLMTimeoutError(`LLM request timed out: ${err.message}`);
      }
      if (err instanceof LLMError) {
        throw err;
      }
      throw new LLMError(`LLM connection failed: ${(err as Error).message}`);
    } finally {
      clearTimeout(timeoutId);
      if (options.signal) {
        options.signal.removeEventListener('abort', onAbort);
      }
    }
  }

  private static validateResponseCode(status: number, bodyText: string): void {
    let errorMessage = bodyText.trim();

    if (errorMessage.length > 0) {
      try {
        const parsed = JSON.parse(errorMessage);
        errorMessage =
          parsed?.error?.message || parsed?.message || errorMessage;
      } catch {
        // Keep original string if not valid JSON
      }
    } else {
      errorMessage = `HTTP Status ${status}`;
    }

    if (status === 401 || status === 403) {
      throw new LLMAuthError(`Authentication failed: ${errorMessage}`);
    } else if (status === 400) {
      if (
        errorMessage.includes('API key not valid') ||
        errorMessage.includes('API_KEY_INVALID')
      ) {
        throw new LLMAuthError(`Authentication failed: ${errorMessage}`);
      } else {
        throw new LLMBadRequestError(`Bad request: ${errorMessage}`);
      }
    } else if (status === 408 || status === 504) {
      throw new LLMTimeoutError(`Request timed out: ${errorMessage}`);
    } else if (status === 429) {
      throw new LLMRateLimitError(`Rate limit exceeded: ${errorMessage}`);
    } else if (status >= 500 && status <= 599) {
      throw new LLMServerError(`Server error (${status}): ${errorMessage}`);
    } else {
      throw new LLMError(`LLM API Error (${status}): ${errorMessage}`);
    }
  }
}
